pkg/logql/sharding.go

```diff
@@ -55,3 +62,255 @@ func (c ConcatLogSelectorExpr) String() string {
 
 	return fmt.Sprintf("%s ++ %s", c.LogSelectorExpr.String(), c.next.String())
 }
+
+type Shards []astmapper.ShardAnnotation
+
+func (xs Shards) Encode() (encoded []string) {
+	for _, shard := range xs {
+		encoded = append(encoded, shard.String())
+	}
+
+	return encoded
+}
+
+// ParseShards parses a list of string encoded shards
+func ParseShards(strs []string) (Shards, error) {
+	if len(strs) == 0 {
+		return nil, nil
+	}
+	shards := make([]astmapper.ShardAnnotation, 0, len(strs))
+
+	for _, str := range strs {
+		shard, err := astmapper.ParseShard(str)
+		if err != nil {
+			return nil, err
+		}
+		shards = append(shards, shard)
+	}
+	return shards, nil
+}
+
+// Downstreamer is an interface for deferring responsibility for query execution.
+// It is decoupled from but consumed by a downStreamEvaluator to dispatch ASTs.
+type Downstreamer interface {
+	Downstream(Expr, Params, Shards) (Query, error)
+}
+
+// DownstreamEvaluator is an evaluator which handles shard aware AST nodes
+type DownstreamEvaluator struct {
+	Downstreamer
+	defaultEvaluator *DefaultEvaluator
+}
+
+func NewDownstreamEvaluator(downstreamer Downstreamer) *DownstreamEvaluator {
+	return &DownstreamEvaluator{
+		Downstreamer: downstreamer,
+		defaultEvaluator: NewDefaultEvaluator(
+			QuerierFunc(func(_ context.Context, p SelectParams) (iter.EntryIterator, error) {
+				// TODO(owen-d): add metric here, this should never happen.
+				return nil, errors.New("Unimplemented")
+			}),
+			0,
+		),
+	}
+}
+
+// Evaluator returns a StepEvaluator for a given SampleExpr
+func (ev *DownstreamEvaluator) StepEvaluator(
+	ctx context.Context,
+	nextEv Evaluator,
+	expr SampleExpr,
+	params Params,
+) (StepEvaluator, error) {
+	switch e := expr.(type) {
+	case DownstreamSampleExpr:
+		// downstream to a querier
+		var shards []astmapper.ShardAnnotation
+		if e.shard != nil {
+			shards = append(shards, *e.shard)
+		}
+		qry, err := ev.Downstream(e.SampleExpr, params, shards)
+		if err != nil {
+			return nil, err
+		}
+
+		res, err := qry.Exec(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return ResultStepEvaluator(res, params)
+
+	case *ConcatSampleExpr:
+		// ensure they all impl the same (SampleExpr, LogSelectorExpr) & concat
+		var xs []StepEvaluator
+		cur := e
+
+		for cur != nil {
+			eval, err := ev.StepEvaluator(ctx, nextEv, cur.SampleExpr, params)
+			if err != nil {
+				// Close previously opened StepEvaluators
+				for _, x := range xs {
+					x.Close()
+				}
+				return nil, err
+			}
+			xs = append(xs, eval)
+			cur = cur.next
+		}
+
+		return ConcatEvaluator(xs)
+
+	case *vectorAggregationExpr, *binOpExpr:
+		return ev.defaultEvaluator.StepEvaluator(ctx, nextEv, e, params)
```

Comment:
Should this be the default ? Or there is a reason for supporting only vectorAggregationExpr and binOpExpr ?

Reply:
Careful I'm not sure I was asking a question.